### PR TITLE
fix: Update the pyarrow to latest v14.0.1 regarding the CVE-2023-47248.

### LINKED ISSUE #3832: Security vulnerability of python package: pyarrow (CVE-2023-47248)
As described by this article regarding CVE-2023-47248: https://securityonline.info/cve-2023-47248-pyarrow-arbitrary-code-execution-vulnerability-a-critical-threat-to-data-analysts/

Current Pyarrow version in setup.py:   "pyarrow>=4,<12",

One conflict from a third-party lib:
1. Snowflake dependency:  pyarrow<10.1.0,>=10.0.1 (from snowflake-connector-python[pandas]==3.4.1->feast (setup.py)). The Snowflake developers are fixing this: https://github.com/snowflakedb/snowflake-connector-python/issues/1802

We need to update the pyarrow version to 14.0.1 

or apply the hotfix:  https://pypi.org/project/pyarrow-hotfix/

### PR DESCRIPTION
**Pull request by shuchu**
**What this PR does / why we need it**:
Update the pyarrow to latest version v14.0.1 which has the fix for CVE-2023-47248

Fixes #3832

### CODE REVIEW

**Comment by shuchu** [comment]
A little bit worried about the unit test coverage.

File: sdk/python/requirements/py3.10-ci-requirements.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #
-# This file is autogenerated by pip-compile with Python 3.10
+# This file is autogenerated by pip-compile with Python 3.9
```
Comment (achals) [inline sdk/python/requirements/py3.10-ci-requirements.txt:2]: I think this is incorrect

Comment (shuchu) [reply sdk/python/requirements/py3.10-ci-requirements.txt:2]: you are right, I need to create a python 3.10 venv and run the command from Makefile. let me fix this.

Comment (shuchu) [reply sdk/python/requirements/py3.10-ci-requirements.txt]: fixed, let's see the testing results.

Comment (shuchu) [reply sdk/python/requirements/py3.10-ci-requirements.txt]: seems integration test failed...

Comment (shuchu) [reply sdk/python/requirements/py3.10-ci-requirements.txt]: 21 fails,  the most frequent error is about the wrong format of Timestamp:    google.api_core.exceptions.BadRequest: 400 Error while reading data, error message: Invalid timestamp microseconds value 1700011424237000000 of logical type NONE; in column 'created' 

let me dig into it and see what's the root cause

Comment (shuchu) [reply sdk/python/requirements/py3.10-ci-requirements.txt]: 1, The Google's BigQuery api only accepts "ms" resolution for timestamp, while the Pyarrow.parquet.write_table() will maintain the resolution to the exact original resolution which is "ns" by default.

**Comment by shuchu** [comment]
I meet a very interesting problem. I only update the Pyarrow version and snowflake api,  the integration test results show me that the timestamp range is error while running Redshift SQL query. 

```
{ error:  Timestamp out of range.\n  
              code:      8001\n  }
```

it happens while run "get_historical_features()"  and the timestamp range were inferenced from the "entity_df":
as in redshift.py::_get_entity_df_event_timestamp_range().
`f"SELECT MIN({entity_df_event_timestamp_col}) AS min, MAX({entity_df_event_timestamp_col}) AS max "`

**Comment by shuchu** [comment]
please do not merge this PR. @sudohainguyen  
It's in a mess status and is for debugging only now.

**Comment by sudohainguyen** [comment]
No worries, looking forward to seeing this works

**Comment by shuchu** [comment]
Finally, I found the fix way. It's about the setting of parameter "coerce_timestamps" of "pyarrow.parquet.write_table".

Let me close this PR and create a clean new one.

**Comment by sudohainguyen** [comment]
Great @shuchu !!

**Pull request closed without merging**